core/market/decentralized/src/matcher.rs

```diff
@@ -140,60 +181,334 @@ impl Matcher {
         demand_id: &SubscriptionId,
         id: &Identity,
     ) -> Result<(), MatcherError> {
-        Ok(self.store.remove_demand(demand_id, id).await?)
+        self.store.remove_demand(demand_id, id).await?;
+
+        log::info!(
+            "Unsubscribed Demand: [{}] using identity: {} [{}]",
+            &demand_id,
+            id.name,
+            id.identity
+        );
+        Ok(())
+    }
+
+    pub async fn list_our_offers(&self) -> Result<Vec<Offer>, QueryOffersError> {
+        let identities = self.identity.list().await?;
+        let store = self.store.clone();
+
+        let mut our_offers = vec![];
+        for node_id in identities.into_iter() {
+            our_offers.append(&mut store.get_offers(Some(node_id)).await?)
+        }
+
+        Ok(our_offers)
+    }
+
+    pub async fn list_our_unsubscribed_offers(
+        &self,
+    ) -> Result<Vec<SubscriptionId>, QueryOffersError> {
+        let identities = self.identity.list().await?;
+        let store = self.store.clone();
+
+        let mut our_offers = vec![];
+        for node_id in identities.into_iter() {
+            our_offers.append(&mut store.get_unsubscribed_offers(Some(node_id)).await?)
+        }
+
+        Ok(our_offers)
     }
 }
 
-pub(crate) async fn on_offer_received(
+// =========================================== //
+// Discovery protocol messages handlers
+// =========================================== //
+
+pub(crate) async fn on_offer_ids_received(
     resolver: Resolver,
     _caller: String,
-    msg: OfferReceived,
-) -> Result<Propagate, ()> {
+    msg: OfferIdsReceived,
+) -> Result<Vec<SubscriptionId>, ()> {
     // We shouldn't propagate Offer, if we already have it in our database.
     // Note that when we broadcast our Offer, it will reach us too, so it concerns
     // not only Offers from other nodes.
-
-    resolver
+    Ok(resolver
         .store
-        .save_offer(msg.offer)
+        .filter_existing(msg.offers)
         .await
-        .map(|offer| {
-            resolver.receive(&offer);
-            Propagate::Yes
-        })
-        .or_else(|e| match e {
-            // Stop propagation for existing, unsubscribed and expired Offers to avoid infinite broadcast.
-            SaveOfferError::Exists(_) => Ok(Propagate::No(Reason::AlreadyExists)),
-            SaveOfferError::Unsubscribed(_) => Ok(Propagate::No(Reason::Unsubscribed)),
-            SaveOfferError::Expired(_) => Ok(Propagate::No(Reason::Expired)),
-            // Below errors are not possible to get from checked_store_offer
-            SaveOfferError::SaveError(_, _)
-            | SaveOfferError::SubscriptionValidation(_)
-            | SaveOfferError::WrongState { .. } => {
-                Ok(Propagate::No(Reason::Error(format!("{}", e))))
+        .map_err(|e| log::warn!("Error filtering Offers. Error: {}", e))?)
+}
+
+pub(crate) async fn on_offers_received(
```

Comment:
same here - maybe?
```suggestion
pub(crate) async fn receive_offers_to_propagate(
```

Reply:
For me your proposal is more misleading. Propagation isn't main goal of this callback, but name suggests it.

Reply:
Ok, my proposition is far from good ;) 

But my point is naming those fns as `on_*` suggests that they are handlers, and they are, but not directly. Their handler nature is depicted via inputting them to the `add_data_handler`.

To be consistent with above I'd rename:
*  `on_offers_received` to `save_and_match_offers` and  
* `filter_known_offer_ids` to `filter_out_known_offer_ids`.
* `on_get_offers` to simply `get_offers`
* `on_offer_unsubscribed` to `unsubscribe_offers`

This will give an idea of what are those fns purposes.
Current names only depicts **how** they are used, and not **what for**.

Reply:
I'm preparing PR based on this PR with my proposals. They are to broad to be contained within comments

Reply:
see #531

Reply:
fixed